hw/formal/tools/jaspergold/parse-fpv-report.py

```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""
+Parses JasperGold FPV logfile and coverage and dump filtered messages in hjson format.
+"""
+
+import argparse
+import re
+from pathlib import Path
+from collections import OrderedDict
+
+import hjson
+
+
+def extract_messages(full_file, patterns, results):
+    """
+    This extracts messages from the sting buffer full_file.
+    The argument patterns needs to be a list of tuples with
+    (<key>, <pattern_to_match_for>).
+    """
+    for severity, pattern in patterns:
+        results[severity] = re.findall(pattern, full_file, flags=re.MULTILINE)
+
+    return results
+
+
+def extract_messages_count(full_file, patterns, results):
+    """
+    This extracts how many times the messages appeared from the sting buffer full_file.
+    The argument patterns needs to be a list of tuples with
+    (<key>, <pattern_to_match_for>).
+    """
+    for severity, pattern in patterns:
+        results[severity] = len(re.findall(pattern, full_file, flags=re.MULTILINE))
+
+    return results
+
+
+def get_results(resdir):
+    """
+    Parse FPV log file and extract errors, warnings, and each property's verification status.
+    """
+    results = OrderedDict()
+    messages = OrderedDict()
+    fpv_summary = OrderedDict()
+    try:
```

Comment:
Since the try block is pretty long, it might be neater to move everything that uses `full_file` into a function. Indeed, I think that then you could return `results` and `fpv_summary`, which makes it a bit clearer what happens with them on an IO error.

Reply:
That looks good. I think there are still some variables that aren't needed: it could now probably be simplified further to something like this:
```python
def get_results(logpath):
    '''Parse FPV log file and extract info to a dictionary'''
    try:
        with Path(logpath).open() as f:
            full_file = f.read()
            results = OrderedDict()
            results['messages'] = parse_fpv_message(full_file)
            fpv_summary = parse_fpv_summary(full_file)
            if fpv_summary:
                results["fpv_summary"] = fpv_summary
            return results

    except IOError as err:
        err_msg = 'IOError {}'.format(err)
        log.error("[get_results] %s", err_msg)
        return {'messages': {'errors': err_msg}}
```

Look, ma, no variables at the top! :-)